garage/logger/snapshotter.py

```diff
@@ -0,0 +1,90 @@
+"""Allows the taking of snapshots."""
+from os import path as osp
+import pickle
+
+
+class Snapshotter:
+    """This class handles the creation of snapshots."""
+
+    def __init__(self):
+        self._snapshot_dir = None
+        self._snapshot_mode = 'all'
+        self._snapshot_gap = 1
+
+    @property
+    def snapshot_dir(self, ):
+        """The directory to save the snapshots in."""
+        return self._snapshot_dir
+
+    @snapshot_dir.setter
+    def snapshot_dir(self, dir_name):
+        self._snapshot_dir = dir_name
+
+    @property
+    def snapshot_mode(self, ):
+        """The type of snapshot to take.
+
+        See save_itr_params.
+        """
+        return self._snapshot_mode
+
+    @snapshot_mode.setter
+    def snapshot_mode(self, mode):
+        self._snapshot_mode = mode
+
+    @property
+    def snapshot_gap(self, ):
+        """The number of iterations between snapshots."""
+        return self._snapshot_gap
+
+    @snapshot_gap.setter
+    def snapshot_gap(self, gap):
+        self._snapshot_gap = gap
+
+    def save_extra_data(self, data, file_name='extra_data.pkl', mode='joblib'):
+        """
+        Data saved here will always override the last entry.
+
+        :param mode: How the dump should be handled.
+        :param file_name: File to dump to.
+        :param data: Something pickle'able.
+        """
+        if self._snapshot_dir:
+            file_name = osp.join(self._snapshot_dir, file_name)
+            if mode == 'joblib':
+                import joblib
+                joblib.dump(data, file_name, compress=3)
+            elif mode == 'pickle':
+                pickle.dump(data, open(file_name, "wb"))
+            else:
+                raise ValueError("Invalid mode: {}".format(mode))
+            return file_name
+
+        return None
+
+    def save_itr_params(self, itr, params):
```

Comment:
API name options:
* `save()`
* `capture()`
* `commit()`
* `put()`
* `record()`
* `write()`

@naeioi @krzentner

Reply:
`save()`, `record()` and `write()` all seem reasonable. `capture()` is weird, since it usually implies **active** observation, and `commit()` implies automicity. `save()` is the one there now, and the top of the list, so let's just use that?

Honestly, this seems like a name change we could land in a separate PR, if it matters.

Reply:
`save()`, `record()` and `write()` all seem reasonable. `capture()` is weird, since it usually implies **active** observation, and `commit()` implies automicity. `save()` is the one there now, and the top of the list, so let's just use that?

Honestly, this seems like a name change we could land in a separate PR, if it matters.